components/brave_federated/util/synthetic_dataset.cc

```diff
@@ -0,0 +1,150 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/util/synthetic_dataset.h"
+
+#include <algorithm>
+#include <cmath>
+#include <fstream>
+#include <random>
+
+#include "base/rand_util.h"
+#include "base/time/time.h"
+
+namespace brave_federated {
+
+SyntheticDataset::SyntheticDataset(std::vector<std::vector<float>> data_points)
+    : data_points_(data_points) {}
+
+SyntheticDataset::SyntheticDataset(std::vector<std::vector<float>> W,
+                                   std::vector<float> b,
+                                   int num_features,
+                                   size_t size) {
+  // Generate time of day and day of week uniformly.
+  const int max_day = 7;
+  const int max_time = 144;
+
+  std::vector<float> cov_x(num_features, 0.0);
+  for (int j = 0; j < num_features; j++) {
+    cov_x[j] = pow((j + 1), -1.2);
+  }
+
+  std::vector<float> mean_x(num_features, 0.0);
+
+  std::vector<std::vector<float>> xs(size,
+                                     std::vector<float>(num_features, 0.0));
+
+  for (size_t j = 0; j < size; j++) {
+    int day_index = base::RandInt(0, max_day);
+    int time_index = base::RandInt(0, max_time);
+
+    xs[j][0] = sin(day_index * 2.0 * M_PI / 7.0);
+    xs[j][1] = cos(day_index * 2.0 * M_PI / 7.0);
+
+    xs[j][2] = sin(time_index * 2.0 * M_PI / 144.0);
+    xs[j][3] = cos(time_index * 2.0 * M_PI / 144.0);
+  }
+
+  std::default_random_engine generator(base::RandDouble());
+
+  for (int i = 4; i < num_features; i++) {
+    std::normal_distribution<float> normal_i(mean_x[i], cov_x[i]);
+    for (size_t j = 0; j < size; j++) {
+      xs[j][i] = normal_i(generator);
+    }
+  }
+
+  std::vector<std::vector<float>> data_points;
+  for (size_t i = 0; i < size; i++) {
+    std::vector<float> tmp = LinearAlgebraUtil::AddVectors(
```

Comment:
Why don't we use `eigen3` directly here? Do we really need a wrapper layer (`LinearAlgebraUtil` + tests)?
Also this make the code slower, because we copy a lot of data during an each operation (for example from std::vector<float> to `eigen3` types and back)

Reply:
Just to clarify, we are referring to 32 parameters of logistic regression. Performance impact should be negligible. I agree it is not ideal, but I am reiterating that we are moving away from Eigen on realistic workloads.